src/loadsave.jl

```diff
@@ -1,37 +1,119 @@
+    path = Base.find_in_path(string(pkg)) # hacky way to determine if a Package is installed
+    path == nothing && return false
+    return isfile(path)
+end
+
+function save_import(pkg::Symbol)
+    !is_installed(pkg)      && throw(NotInstalledError(pkg, ""))
+    !isdefined(Main, pkg)   && eval(Main, Expr(:import, pkg))
+    return Main.(pkg)
+end
+
+
+for (applicable_, add_, dict_) in (
+        (:applicable_loaders, :add_loader, :sym2loader),
+        (:applicable_savers,  :add_saver,  :sym2saver))
     @eval begin
-        $appl{sym}(::Formatted{DataFormat{sym}}) = get($dct, sym, [:FileIO]) # if no loader is declared, fallback to FileIO
-        function $fchk(pkg::Symbol)
-            !isdefined(Main, pkg) && eval(Main, Expr(:import, pkg))
-            return Main.(pkg)
-        end
-        function $fadd{sym}(::Type{DataFormat{sym}}, pkg::Symbol)
-            list = get($dct, sym, Symbol[])
-            $dct[sym] = push!(list, pkg)
+        $applicable_{sym}(::@compat(Union{Type{DataFormat{sym}}, Formatted{DataFormat{sym}}})) = get($dict_, sym, [:FileIO]) # if no loader is declared, fallback to FileIO
+        function $add_{sym}(::Type{DataFormat{sym}}, pkg::Symbol)
+            list = get($dict_, sym, Symbol[])
+            $dict_[sym] = push!(list, pkg)
         end
     end
 end
 
-applicable_mime{sym}(::MIME{sym}) = get(mimedict, sym, [:nothing])
-function check_mime(pkg::Symbol)
-    pkg == :nothing && error("No MIME package available")
-    !isdefined(Main, pkg) && eval(Main, Expr(:import, pkg))
-    return pkg
-end
+applicable_mime{sym}(::MIME{sym}) = get(mimedict, sym, [(Any, :FileIO)])
 
 @doc """
 `add_mime(mime, T, :Package)`  triggers `using Package` before attempting to write object of type `T` in format `mime`.
 """ ->
 function add_mime{sym,T}(::MIME{sym}, ::Type{T}, pkg::Symbol)
-    list = get(mimedict, sym, Any[])
+    list = get(mimedict, sym, @compat(Tuple{DataType, Symbol})[])
     mimedict[sym] = push!(list, (T,pkg))
 end
 
 @doc "`add_loader(fmt, :Package)` triggers `using Package` before loading format `fmt`" -> add_loader
 @doc "`add_saver(fmt, :Package)` triggers `using Package` before saving format `fmt`" -> add_saver
+
+
+@doc """
+- `load(filename)` loads the contents of a formatted file, trying to infer
+the format from `filename` and/or magic bytes in the file.
+- `load(strm)` loads from an `IOStream` or similar object. In this case,
+the magic bytes are essential.
+- `load(File(format"PNG",filename))` specifies the format directly, and bypasses inference.
+- `load(f; options...)` passes keyword arguments on to the loader.
+""" ->
+function load(s::@compat(Union{AbstractString,IO}), args...; options...)
+    q           = query(s)
+    libraries   = applicable_loaders(q)
+    failures    = Any[]
+    for library in libraries
+        try
+            Library = save_import(library)
+            return Library.load(q, args...; options...)
```

Comment:
I'm not getting what's so bad about the hybrid solution. It seems elegant to me as it gives us the best from two worlds :D 
If we don't export, we could also just do `load(::Formatted)` in `ImageMagick`, solving your second concerns.
I think we shouldn't worry about the compilation for the method, since loading different types requires different methods. With multiple dispatch, we have the advantage to call them all the same, while they're logically different functions. So having LLVM emit different functions for different formats is all well and good in my view.
In contrast, I consider it as bad design, that all formats are meshed up in one function like in `ImageMagick`, but it's just how it is when you have one monolithic c-library handling everything for you. So it's fine in that case and I think we have everything at our hands to handle this case well. I don't see why this one pathological case is an argument for dropping the whole dispatch design.